src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/HttpClientFactoryDisposeTests.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace Microsoft.Extensions.Http.Tests
+{
+    public class HttpClientFactoryDisposeTests
+    {
+        [Fact]
+        public async Task DisposingServiceProvider_DisposesHttpClientFactory_ReleasesResources()
+        {
+            // Arrange
+            var disposeCounter = new DisposeCounter();
+            var services = new ServiceCollection();
+            services.AddSingleton(disposeCounter);
+            services.AddTransient<DisposeTrackingHandler>();
+
+            services.AddHttpClient("test")
+                .AddHttpMessageHandler<DisposeTrackingHandler>()
+                .SetHandlerLifetime(TimeSpan.FromMilliseconds(50)); // Very short to ensure quick expiration
+
+            var serviceProvider = services.BuildServiceProvider();
+            var factory = serviceProvider.GetRequiredService<IHttpClientFactory>();
+
+            // Act - create and use clients
+            for (int i = 0; i < 5; i++)
+            {
+                using var client = factory.CreateClient("test");
+                try
+                {
+                    using var cts = new CancellationTokenSource(millisecondsDelay: 1);
+                    await client.GetAsync("http://example.com/", cts.Token)
+                        .ContinueWith(t => { }); // Ignore errors
+                }
+                catch
+                {
+                    // Ignore errors
+                }
+            }
+
+            // Wait for handlers to expire
```

Comment:
if we imagine that we somehow ensured all the handlers made their ways to the expired handlers collection, then such a test will only test half of the introduced logic

we also need to test that all handlers from _active_ handlers are cleared up.